### PR TITLE
HPCC-22395 Fix Dali WhiteList (make sure using peer IP)

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

**Comment by jakesmith** [comment]
NB: This change stores the peer address of the socket connection along with the channel (CMPChannel) created. When a session registration happens, it looks up the channel and retrieves the peer to pass to the whitelist.

The MP mechanism has by this stage already connected (socket, and MP channel), serialized the client IP (twice in fact), received a MP message, queued and dequeued.

It would have been preferable and more general to reject at connection time, and the whitelist mechanism could later be moved to either the communicator level or even down to the select handler, with an interface/callback to checking incoming connections.
However, that would make it more difficult to deserialize criteria to reject on (e.g. role) and to pass back meaningful errors, as you don't know the protocol at that stage.

File: system/mp/mpcomm.cpp [hunk 1]
```diff
@@ -2250,7 +2256,7 @@ bool CMPServer::recv(CMessageBuffer &mbuf, const SocketEndpoint *ep, mptag_t tag
         bool notify(CMessageBuffer *msg)
         {
             if ((tag==TAG_ALL)||(tag==msg->getTag())) {
-                SocketEndpoint senderep = msg->getSender();
+                const SocketEndpoint &senderep = msg->getSender();
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp]: NB: this is an incidental change (just avoiding a copy).

**Comment by jakesmith** [comment]
@ghalliday - please also review.

**Comment by ghalliday** [comment]
@richardkchapman it looks correct to me.

File: dali/base/dacoven.cpp [hunk 2]
```diff
@@ -321,6 +321,12 @@ class CCovenBase: implements ICoven, public CInterface
         if (comm)
             comm->cancel(srcrank,tag);
     }
+    virtual const SocketEndpoint &queryChannelPeerEndpoint(const SocketEndpoint &sender) override
+    {
+        if (!comm)
```
Comment (richardkchapman) [inline dali/base/dacoven.cpp]: assertex(comm) would be clearer, perhaps (and give a marginally more helpful error if triggered).

**Comment by richardkchapman** [comment]
@jakesmith A couple of trivial comments - I am happy to merge as-is if you don't feel like fixing them

File: dali/base/dacoven.cpp (hunk 2 above)
Comment (jakesmith) [reply dali/base/dacoven.cpp]: Will change this and a similar one in mpcomm.cpp's queryChannelPeerEndpoint() to an assertex

**Comment by jakesmith** [comment]
@richardkchapman - have changed the two introduced throwUnexpected()'s to assertex's

**Pull request merged**